lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+        # from a sweepline intersector.
+        #
+        # Can be used to determine which edges need to be split
+        # after adding edges.
+        def intersection_map
+          intersector = SweeplineIntersector.new(edges)
+          intersections = intersector.proper_intersections
+
+          intersection_map = {}
+          intersections.each do |int|
+            # check the int_point against each edge.
+            # if it is not on the boundary of the edge, add it to the
+            # list of intersections for that edge.
+            s1_intersects = int.point != int.s1.s && int.point != int.s1.e
+            if s1_intersects
+              unless intersection_map[int.s1]
+                intersection_map[int.s1] = []
+              end
+              intersection_map[int.s1] << int.point
+            end
+
+            s2_intersects = int.point != int.s2.s && int.point != int.s2.e
+            next unless s2_intersects
+            unless intersection_map[int.s2]
+              intersection_map[int.s2] = []
+            end
+            intersection_map[int.s2] << int.point
+          end
+          intersection_map
+        end
+
+        def create_half_edges
+          @edges.each do |edge|
+            create_half_edge(edge)
+          end
+        end
+
+        def create_half_edge(edge)
+          e1, e2 = HalfEdge.from_edge(edge)
+          insert_half_edge(e1)
+          insert_half_edge(e2)
+        end
+
+        def insert_half_edge(he)
+          unless incident_edges[he.origin.coordinates]
+            @incident_edges[he.origin.coordinates] = []
+          end
+          @incident_edges[he.origin.coordinates] << he
+        end
+
+        # Links all half-edges where possible.
+        # Defines +next+ and +prev+ for every half-edge by rotating
+        # through all half-edges originating at a vertex.
+        #
+        # Assuming half-edges are sorted CCW, every sequential pair of
+        # half-edges (e1, e2) can be linked by saying e1.prev = e2.twin
+        # and e2.twin.next = e1.
+        def link_half_edges
+          incident_edges.each_value do |hedges|
+            hedges.sort!
+            if hedges.size > 1
+              (0..hedges.size - 2).each do |i|
+                e1 = hedges[i]
+                e2 = hedges[i + 1]
+
+                e1.prev = e2.twin
+                e2.twin.next = e1
+              end
+
+              hedges[-1].prev = hedges[0].twin
+              hedges[0].twin.next = hedges[-1]
+
+            # handle case of dangling line
+            else
+              e = hedges[0]
+              e.twin.next = e
+              e.prev = e.twin
+            end
+          end
+        end
+
+        # It is possible that intersections occur when new edges are added.
+        # This will split those edges into more half-edges while preserving
+        # the existing half-edges when possible since geometries may reference
+        # them.
```

Comment:
Not entirely sure I get the underlying issue there..

Reply:
Say we are making a `GeometryGraph` with the following geometry `GEOMETRYCOLLECTION(POLYGON((0 0,0 1,1 1,1 0,0 0)),LINESTRING(0.5 0.5,0.5 -0.5))`.
![image](https://user-images.githubusercontent.com/12189611/125644136-76a8ee4d-4ecd-47b9-9874-1ad00c3663b0.png)

We add each component of the collection separately, so first we will add the polygon and create the half-edges (only showing the bottom).

![image](https://user-images.githubusercontent.com/12189611/125644292-c1662edc-d000-4142-a85b-590439c61108.png)

The `GeometryGraph` will reference the half-edge on the interior of the polygon in its `geom_edges`. This is how it can access a half-edge on the interior of the polygon for future uses.

Now, when we go to overlay the next component, we need to split those half-edges since there's an intersection.
![image](https://user-images.githubusercontent.com/12189611/125640305-1b7c2e6d-f895-4e09-8ec1-6d17ba1308f4.png)

But it's important that we don't just remove the existing half-edges and replace them (although that would be easier), since the `GeometryGraph` will hold a reference to the red half-edge at the bottom and if we replace it then the `GeometryGraph` will no longer be able to properly access the graph.

![image](https://user-images.githubusercontent.com/12189611/125643957-a89a0f35-2e5b-4a52-be87-aedb9463b901.png)

So when we split the edges from intersections we need to preserve the existing half-edges and change their `twin` and `next` values.